### PR TITLE
[SPARK-47300][SQL] `quoteIfNeeded` should quote identifier starts with digits

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

`quoteIfNeeded` is used to generate pretty strings of identifiers in error message, EXPLAIN result, etc. It's mostly for humans to read but sometimes people may copy-paste the identifier string and put it in a SQL. There is a small issue in `quoteIfNeeded`: it does not quote number literals like `0d`, and people will get parse exception if they directly put the identifier string in SQL.

This PR fixes the issue by always quoting the identifier if it does not start with alphabet or underscore. Note, there might be program trying to parse the identifier string, but this change is safe as these programs should already handle quoting and it's ok to quote more.

### Why are the changes needed?

make identifier string parsable is more user-friendly.

### Does this PR introduce _any_ user-facing change?

yes, the identifier string in the error message and EXPLAIN result will be properly quoted.

### How was this patch tested?

new test suite

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @yaooqinn

**Comment by yaooqinn** [comment]
```
spark-sql (default)> select version();
3.5.0 ce5ddad990373636e94071e7cef2f31021add07b
spark-sql (default)> create table d (0a real) using parquet;
spark-sql (default)> desc formatted d;
0a                  	float
```

Do we already have it?

**Comment by cloud-fan** [comment]
@yaooqinn `0a` is fine but `0d` is not, as it's double literal.

**Review by yaooqinn — APPROVED** [review]
LGTM

**Review by HyukjinKwon — APPROVED** [review]

**Review by MaxGekk — COMMENTED** [review]
@cloud-fan Seems the test failure is related:
```
[info] - SPARK-34872: quoteIfNeeded should quote a string which contains non-word characters *** FAILED *** (3 milliseconds)
[info]   "[`1a`]" did not equal "[1a]" (StringUtilsSuite.scala:136)
[info]   Analysis:
[info]   "[`1a`]" -> "[1a]"
```

**Comment by MaxGekk** [comment]
How about deduplicate the tests:
https://github.com/apache/spark/blob/5089140e2e6a43ffef584b42aed5cd9bc11268b6/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/StringUtilsSuite.scala#L132-L140
and maybe move them to `QuotingUtilsSuite`.

**Comment by dongjoon-hyun** [comment]
It seems that the same failure (which @MaxGekk 's reported) still exists at the last CI.
```
StringUtilsSuite:
[info] - escapeLikeRegex (1 millisecond)
[info] - filter pattern (1 millisecond)
[info] - string concatenation (1 millisecond)
[info] - string concatenation with limit (1 millisecond)
[info] - string concatenation return value (1 millisecond)
[info] - SPARK-31916: StringConcat doesn't overflow on many inputs (190 milliseconds)
[info] - SPARK-31916: verify that PlanStringConcat's output shows the actual length of the plan (3 milliseconds)
[info] - SPARK-34872: quoteIfNeeded should quote a string which contains non-word characters *** FAILED *** (2 milliseconds)
[info]   "[`1a`]" did not equal "[1a]" (StringUtilsSuite.scala:136)
[info]   Analysis:
[info]   "[`1a`]" -> "[1a]"
```

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (if CIs passes after adjusting the test suite)

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/StringUtilsSuite.scala [hunk 1]
```diff
@@ -129,16 +129,6 @@ class StringUtilsSuite extends SparkFunSuite with SQLHelper {
     }
   }
 
-  test("SPARK-34872: quoteIfNeeded should quote a string which contains non-word characters") {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/StringUtilsSuite.scala:132]: it's covered by the new test

**Review by MaxGekk — APPROVED** [review]

**Review by yaooqinn — APPROVED** [review]

**Pull request closed without merging**